flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/NetworkInput.java

```diff
@@ -0,0 +1,260 @@
+import org.apache.flink.runtime.io.network.api.serialization.RecordDeserializer.DeserializationResult;
+import org.apache.flink.runtime.io.network.api.serialization.SpillingAdaptiveSpanningRecordDeserializer;
+import org.apache.flink.runtime.io.network.buffer.Buffer;
+import org.apache.flink.runtime.io.network.partition.consumer.BufferOrEvent;
+import org.apache.flink.runtime.io.network.partition.consumer.InputGate;
+import org.apache.flink.runtime.io.network.partition.consumer.InputGateListener;
+import org.apache.flink.runtime.plugable.DeserializationDelegate;
+import org.apache.flink.runtime.plugable.NonReusingDeserializationDelegate;
+import org.apache.flink.streaming.api.watermark.Watermark;
+import org.apache.flink.streaming.runtime.streamrecord.StreamElement;
+import org.apache.flink.streaming.runtime.streamrecord.StreamElementSerializer;
+import org.apache.flink.streaming.runtime.streamstatus.StatusWatermarkValve;
+import org.apache.flink.streaming.runtime.streamstatus.StreamStatus;
+
+import java.io.IOException;
+import java.util.BitSet;
+import java.util.Collection;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+import static org.apache.flink.util.Preconditions.checkState;
+
+/**
+ * Implementation of taking {@link InputGate} as {@link Input}.
+ */
+@Internal
+public final class NetworkInput implements Input, InputGateListener {
+
+	private final int inputId;
+
+	private final InputGate inputGate;
+
+	private final DeserializationDelegate<StreamElement> deserializationDelegate;
+
+	private final RecordDeserializer<DeserializationDelegate<StreamElement>>[] recordDeserializers;
+
+	/**
+	 * Valves that control how watermarks and stream statuses from the input are forwarded.
+	 */
+	private final StatusWatermarkValve statusWatermarkValve;
+
+	private final StreamStatusHandler streamStatusHandler;
+
+	/** Registered listener to forward input notifications to. */
+	private volatile CompletableFuture<?> inputListener = null;
+
+	/**
+	 * The channel from which a buffer came, tracked so that we can appropriately map
+	 * the watermarks and watermark statuses to the correct channel index of the correct valve.
+	 */
+	private int currentChannel = -1;
+
+	private RecordDeserializer<DeserializationDelegate<StreamElement>> currentRecordDeserializer = null;
+
+	/** The new overall watermark aggregated on all aligned input channels. */
+	private Watermark alignedWatermark = null;
+
+	private final BitSet finishedChannels;
+
+	private boolean isFinished = false;
+
+	public NetworkInput(
+		int inputId,
+		Collection<InputGate> inputGates,
+		TypeSerializer<?> inputSerializer,
+		StreamStatusHandler streamStatusHandler,
+		IOManager ioManager) {
+
+		this.inputId = inputId;
+
+		this.inputGate = InputGateUtil.createInputGate(inputGates.toArray(new InputGate[inputGates.size()]));
```

Comment:
It is recommended to pass `new InputGate[0]` since ~Java1.6, the call is using intrinics.